drivers/iio/common/st_sensors/st_sensors_core.c

```diff
@@ -684,6 +689,76 @@ ssize_t st_sensors_sysfs_scale_avail(struct device *dev,
 }
 EXPORT_SYMBOL(st_sensors_sysfs_scale_avail);
 
+/*
+ * st_sensors_sysfs_show_always_on() - get the value of the always_on flag.
+ *
+ * @dev: device reference.
+ * @attr: device attribute.
+ * @buf: sysfs buffer.
+ *
+ * Return: Number of bytes printed into the buffer
+ */
+ssize_t st_sensors_sysfs_show_always_on(struct device *dev,
+					struct device_attribute *attr,
+					char *buf)
+{
+	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
+	struct st_sensor_data *sdata = iio_priv(indio_dev);
+
+	return scnprintf(buf, PAGE_SIZE, "%d\n", sdata->always_on);
+}
+EXPORT_SYMBOL(st_sensors_sysfs_show_always_on);
+
+/*
+ * st_sensors_sysfs_store_always_on() - set/unset always_on flag.
+ *				       Accepted values are:
+ *				       - 1: to set the flag and keep the
+ *					    device always enabled.
+ *				       - 0: to unset the flag and enable the
+ *					    device just during data access.
+ *
+ * @dev: device reference.
+ * @attr: device attribute.
+ * @buf: sysfs buffer.
+ * @count: number of bytes used from the buffer.
+ *
+ * Return: Either the number of bytes used from the buffer or an error code.
+ */
+ssize_t st_sensors_sysfs_store_always_on(struct device *dev,
+					 struct device_attribute *attr,
+					 const char *buf, size_t count)
+{
+	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
+	struct st_sensor_data *sdata = iio_priv(indio_dev);
+	unsigned int val;
+	int ret;
+
+	ret = kstrtouint(buf, 10, &val);
+	if (ret < 0)
+		return ret;
+
+	if (val != 0 && val != 1)
+		return -EINVAL;
+
```

Comment:
Would `mutex_lock(&indio_dev->mlock);` make sense here? Or is this function called with mutex already acquired?
I'm thinking about scenario in which userspace would accesses both raw attribute and always_on in parallel, e.g. disabling always_on in middle of raw read.

Reply:
As `always_on` is only read after reading the raw value in `st_sensors_read_info_raw` and the only write to it in `st_sensors_sysfs_store_always_on` should be atomic, this shouldn't be a problem.

Reply:
And if `always_on` is written 0 while raw read executes
```
err = st_sensors_read_axis_data(indio_dev, ch, val);
```
?
Couldn't the sensor be then disabled in the middle of reading?

By the way, Massimo (@max186), are you going to follow the advice from Jonathan Cameron and try to implement the approach using PM framework? If yes, then it doesn't make sense to discuss here further ;)

Reply:
I already discussed with @zandrey about Jonathan's suggestions and we considered more suitable going for the buffered interface solution. So, yes, it doesn't make sense to continue this discussion :)